lib/rex/proto/ms_dtyp.rb

```diff
@@ -0,0 +1,334 @@
+    class MsDtypSystemAuditObjectAce < BinData::Record
+      endian :little
+
+      ms_dtyp_ace_header                   :header, initial_value: { ace_type: 7 }
+      ms_dtyp_system_audit_object_ace_body :body
+    end
+
+    class MsDtypAce < BinData::Record
+      endian :little
+  
+      ms_dtyp_ace_header :header
+      choice             :body, selection: -> { header.ace_type } do
+        ms_dtyp_access_allowed_ace_body 0
+        ms_dtyp_access_denied_ace_body 1
+        ms_dtyp_system_audit_ace_body 2
+        # Type 3 is reserved for future use
+        # Type 4 is reserved for future use
+        ms_dtyp_access_allowed_object_ace_body 5
+        ms_dtyp_access_denied_object_ace_body 6
+        #ms_dtyp_system_audit_object_ace_body 7
+        # Type 8 is reserved for future use
+        # Type 14 aka 0xE is reserved for future use
+        # Type 16 aka 0x10 is reserved for future use
+        string                          :default, read_length: -> { header.ace_size - body.rel_offset }
+      end
+    end
+  
+    # [2.4.5 ACL](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-dtyp/20233ed8-a6c6-4097-aafa-dd545ed24428)
+    class MsDtypAcl < BinData::Record
+      ACL_REVISION = 2
+      ACL_REVISION_DS = 4
+  
+      endian :little
+  
+      uint8  :acl_revision, initial_value: ACL_REVISION
+      uint8  :sbz1
+      uint16 :acl_size, initial_value: -> { num_bytes }
+      uint16 :acl_count, initial_value: -> { aces.length }
+      uint16 :sbz2
+      array  :aces, type: :ms_dtyp_ace, initial_length: :acl_count
+    end
+  
+    # [2.4.6 SECURITY_DESCRIPTOR](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-dtyp/7d4dac05-9cef-4563-a058-f108abecce1d)
+    class MsDtypSecurityDescriptor < BinData::Record
+      endian :little
+  
+      uint8  :revision, initial_value: 1
+      uint8  :sbz1
+      struct :control do
+        bit1 :ss
+        bit1 :dt
+        bit1 :sd
+        bit1 :sp, initial_value: -> { sacl? ? 1 : 0 }
+        bit1 :dd
+        bit1 :dp, initial_value: -> { dacl? ? 1 : 0 }
+        bit1 :gd
+        bit1 :od
+  
+        bit1 :sr, initial_value: 1
+        bit1 :rm
+        bit1 :ps
+        bit1 :pd
+        bit1 :si
+        bit1 :di
+        bit1 :sc
+        bit1 :dc
+      end
+      uint32 :offset_owner, value: -> { offset_for(:owner_sid) }
+      uint32 :offset_group, value: -> { offset_for(:group_sid) }
+      uint32 :offset_sacl, value: -> { offset_for(:sacl) }
+      uint32 :offset_dacl, value: -> { offset_for(:dacl) }
+      rest   :buffer, value: -> { build_buffer }
+      hide   :buffer
+  
+      def initialize_shared_instance
+        # define accessor methods for the custom fields to expose the same API as BinData
+        define_field_accessors_for2(:owner_sid)
+        define_field_accessors_for2(:group_sid)
+        define_field_accessors_for2(:sacl)
+        define_field_accessors_for2(:dacl)
```

Comment:
This isn't my code and is instead Spencer's code but I believe the idea here was to define something like `.group_sid?` and then have that return True or False based on if it was present, whereas `attr_accessor` would define the read and write methods for reading and writing these options.

Reply:
As per our slack conversation and the above comment, closing this out in favor of addressing https://github.com/rapid7/metasploit-framework/pull/17122#discussion_r1000377775 and renaming the method since both the `attr_accessor` and the `define_field_accessors_for2` method are needed, its just that `define_field_accessors_for2` actually does something different than its name implies.